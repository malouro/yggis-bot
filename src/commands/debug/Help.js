import { Command } from '../../classes'

export default class Help extends Command {
	constructor() {
		super({
			name: 'Help',
			category: 'debug',
			aliases: [],
			description: 'Help menus & support',
			usage: {
				name: 'help',
				args: [
					{
						name: 'command',
						description: 'Get help on how to use a specific command.',
						chainable: false,
					},
					{
						name: 'category',
						description: 'Learn more about commands available under a given category.',
						chainable: false,
					},
				],
			},
		})

		this.messageOutput = ''
	}

	// TODO
	// eslint-disable-next-line no-unused-vars
	buildMainHelpMenu(config, commands, categories) {
		return 'This is the main help menu.'
	}

	buildCommandHelpMenu(config, command, invalidUsage = false) {
		const { description, usage } = command

		const chainableArgs = []
		const nonChainableArgs = []

		usage.args.forEach((arg) => {
			if (arg.chainable) {
				chainableArgs.push(arg.name)
			} else {
				nonChainableArgs.push(arg.name)
			}
		})

		const getArgsUsage = () => {
			let argsUsageOutput = ''

			if (nonChainableArgs.length > 0) {
				argsUsageOutput = argsUsageOutput.concat(
					'`<',
					nonChainableArgs.join('|'),
					'>`',
				)
			}

			if (chainableArgs.length > 0) {
				argsUsageOutput = argsUsageOutput.concat(
					'`(',
					chainableArgs.join(', '),
					')`',
				)
			}

			return argsUsageOutput
		}

		const getDetailedArgsDescription = () => {
			let argsDetailsOutput = ''

			usage.args.forEach((arg, index) => {
				if (arg.description) {
					argsDetailsOutput = argsDetailsOutput.concat(
						index === 0 ? '\n' : '\n\n',
						`\`${arg.name}\``,
						'\n',
						arg.description,
						arg.usage ? `${'\n'.repeat(2)}\`${arg.usage}\`` : '',
					)
				}
			})

			return argsDetailsOutput
		}

		const title = `**${command.name} Command**`
		const commandUsage = `\`${config.commandPrefix}${usage.name}\` ${getArgsUsage()}`

		return [
			invalidUsage
				? `Invalid \`${config.commandPrefix}${this.name}\` usage.`
				: title,
			'',
			description,
			'',
			commandUsage,
			getDetailedArgsDescription(),
		].join('\n')
	}

	// TODO
	buildCategoryHelpMenu() {
		return 'This is the command category help menu'
	}

	preAction({ args, bot }) {
		if (args.length === 1) {
			// !help (no args)
			this.messageOutput = this.buildMainHelpMenu(
				bot.config,
				bot.commands,
				bot.commandCategories,
			)
		} else if (bot.commands.has(args[1])) {
			// !help <command>
			this.messageOutput = this.buildCommandHelpMenu(
				bot.config,
				bot.commands.get(args[1].toLocaleLowerCase()),
			)
		} else if (bot.commandCategories.has(args[1])) {
			// !help <category>
			this.messageOutput = this.buildCategoryHelpMenu(
				bot.config,
				bot.commands.get(args[1].toLocaleLowerCase()),
			)
		} else {
			// !help {invalidUsage}
			this.messageOutput = this.buildCommandHelpMenu(bot.config, this, true)
		}
	}

	action({ message }) {
		message.channel.send(this.messageOutput)
	}
}
