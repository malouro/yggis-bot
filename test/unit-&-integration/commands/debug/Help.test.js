import { runCommand, makeMockBot } from '../../../testHelpers'

import Help from '../../../../src/commands/debug/Help'

describe('Help Command', () => {
	const HelpCommand = new Help()
	const bot = makeMockBot({ mockCommand: Help })

	describe('Main help menu', () => {
		test('returns the main help menu', async () => {
			const {
				message: {
					channel: { send: mockedSend },
				},
			} = await runCommand(HelpCommand, { args: ['!help'] })
			const mainHelpMenu = mockedSend.mock.results[0].value

			expect(mainHelpMenu).toMatchSnapshot()
		})
	})

	describe('Help Command help menu', () => {
		let mockedSend = null
		let helpCommandHelpMenu = null

		beforeEach(async () => {
			({
				message: {
					channel: { send: mockedSend },
				},
			} = await runCommand(HelpCommand, {
				args: ['!help', 'help'],
				bot,
			}))

			helpCommandHelpMenu = mockedSend.mock.results[0].value
		})

		test('returns the command\'s help menu', async () => {
			expect(helpCommandHelpMenu).toMatchSnapshot()
		})

		test.each(
			HelpCommand.usage.args.map(
				arg => [arg.name, arg.description]
			),
		)(
			'the help command\'s help menu shows the `%s` argument\'s description',
			async (argName, argDescription = '') => {
				expect(helpCommandHelpMenu).toEqual(expect.stringContaining(argDescription))
			},
		)
	})

	describe('Command help menu', () => {
		test.todo(
			'displays chainable and non-chainable arguments in the command help menu'
		)
		test.todo('displays a command\'s argument usage when available')
	})

	describe('Command category help menu', () => {
		test('returns a category\'s help menu', async () => {
			const {
				message: {
					channel: { send: mockedSend },
				},
			} = await runCommand(HelpCommand, {
				args: ['!help', 'debug'],
				bot,
			})
			const debugCategoryHelpMenu = mockedSend.mock.results[0].value

			expect(debugCategoryHelpMenu).toMatchSnapshot()
		})
	})
})
