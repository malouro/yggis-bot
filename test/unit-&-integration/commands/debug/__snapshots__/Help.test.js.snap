// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Help Command Command category help menu returns a category's help menu 1`] = `"This is the command category help menu"`;

exports[`Help Command Help Command help menu returns the command's help menu 1`] = `
"**Help Command**

Help menus & support

\`!help\` \`<command|category>\`

\`command\`
Get help on how to use a specific command.

\`category\`
Learn more about commands available under a given category."
`;

exports[`Help Command Main help menu returns the main help menu 1`] = `"This is the main help menu."`;
